docs/2018-11-12-Unification.md

```diff
@@ -0,0 +1,360 @@
+Examples include patterns having constructors at the top and domain values.
+
+### Constructor vs sort injection
+
+```
+Constr(φ1, φ2, ..., φn) ∧ inj(s2, s1)(ψ) = ⊥
+inj(s2, s1)(ψ) ∧ Constr(φ1, φ2, ..., φn) = ⊥
+
+not-handled-error otherwise
+```
+
+### Different constructors
+
+Does not actually test for the distinctness part, assumes that the
+"equal injective heads" case was checked before
+
+```
+Constr1(φ1, φ2, ..., φn) ∧ Constr2(ψ1, ψ2, ..., ψn) = ⊥
+
+not-handled-error otherwise
+```
+
+### Builtin map unification
+
+Note that currently map keys cannot have variables, so it does not make sense
+to worry about alpha equivalence.
+
+```
+{k1 -> φ1, k2 -> φ2, ..., km -> φm} ∧ {l1 -> ψ1, l2 -> ψ2, ..., ln -> ψn}
+  = ⊥
+    if {k1, k2, ..., km} ≠ {l1, l2, ..., ln}
+  = {k1 -> φ1 ∧ ψ1, k2 -> φ2 ∧ ψ2, ..., kn -> φn ∧ ψn}
+    otherwise
+
+{k1 -> φ1, k2 -> φ2, ..., km -> φm}
+    ∧ concat({l1 -> ψ1, l2 -> ψ2, ..., ln -> ψn}, x)
+  = ⊥
+    if {l1, l2, ..., ln} ⊄ {k1, k2, ..., kn}
+  = {k1 -> φ1 ∧ ψ1, k2 -> φ2 ∧ ψ2, ..., km -> φm ∧ ψm}
+    ∧ x = {lm+1 -> ψm+1, ..., ln -> ψn}
+    renumbering such that {l1, l2, ..., lm} = {k1, k2, ..., km}
+
+{k1 -> φ1, k2 -> φ2, ..., km -> φm}
+    ∧ concat(x, {l1 -> ψ1, l2 -> ψ2, ..., ln -> ψn})
+  = identical to the above
+
+dv ∧ concat(ψ1, ψ2)
+  = dv ∧ ⌈dv ∧ concat(ψ1, ψ2)⌉   -- ** different from the actual implementation.
+    where dv = {k1 -> φ1, k2 -> φ2, ..., km -> φm}
+
+{k1 -> φ1, k2 -> φ2, ..., km -> φm} ∧ element(l, ψ)
+  = ⊥                     if m > 1
+  = {k1 ∧ l -> φ1 ∧ ψ}    otherwise
+
+ψ ∧ {k1 -> φ1, k2 -> φ2, ..., km -> φm}
+  = {k1 -> φ1, k2 -> φ2, ..., km -> φm} ∧ ψ
+
+not-handled-error otherwise
+```
+
+The evaluation of `{k1 -> φ1 ∧ ψ1, k2 -> φ2 ∧ ψ2, ..., kn -> φn ∧ ψn}`
+is similar to the evaluation for equal injective heads and goes like this:
+
+```
+evaluate each φi ∧ ψi by calling unify(φi, ψi)
+if the result is an error, return handled-with-error
+if the result is ⊥, return ⊥  -- actually handled through the general
+                              -- case below.
+Otherwise, represent the result as ti ∧ si ∧ pi
+
+return {k1 -> t1, k2 -> t2, ..., kn -> tn}
+       ∧ (s1 ∧ s2 ∧ ... ∧ sn)
+       ∧ (p1 ∧ p2 ∧ ... ∧ pn)
+```
+
+The evaluation of `{k1 ∧ l -> φ1 ∧ ψ}` is similar.
+
+### Builtin set unification
+
+Note that currently set elements cannot have variables, so it does not
+make sense to worry about alpha equivalence.
+
+```
+{k1, k2, ..., km} ∧ {l1, l2, ..., ln}
+  = ⊥                    if {k1, k2, ..., km} ≠ {l1, l2, ..., ln}
+  = {k1, k2, ..., kn}    otherwise
+
+{k1, k2, ..., km} ∧ concat({l1, l2, ..., ln}, x)
+  = ⊥
+    if {l1, l2, ..., ln} ⊄ {k1, k2, ..., km}
+  = {k1, k2, ..., kn} ∧ x = {kn+1, ..., km}
+    where elements are renumbered such that {l1, ..., ln} = {k1, ..., kn}
+
+{k1, k2, ..., km} ∧ concat(x, {l1, l2, ..., ln})
+  = same as above
+
+dv ∧ concat(ψ1, ψ2)
+  = dv ∧ ⌈dv ∧ concat(ψ1, ψ2)⌉   -- ** different from the actual implementation.
```

Comment:
I think, but I am not at all sure, that the unification code will return `⌈dv ∧ concat(ψ1, ψ2)⌉ ` instead of `dv ∧ ⌈dv ∧ concat(ψ1, ψ2)⌉ `, hopefully @ttuegel can tell us whether this is the case or not. Otherwise I'll add a test for this specific behaviour (I didn't find any testing it).

Reply:
As for Map, the implementation returns
```suggestion
  = dv ∧ concat(ψ1, ψ2) ∧ ⌈dv ∧ concat(ψ1, ψ2)⌉
```
although it makes sense to use what you have written instead.

Reply:
Ah, ok, I'd change the implementation then.